### PR TITLE
feat(tts): integrate LavaSR audio enhancer as opt-in post-processing

### PR DESCRIPTION
**Pull request by sharmaraju352**
## 🎯 What problem does this PR solve?

- TTS addon outputs audio at native engine sample rates (Chatterbox 24kHz, Supertonic 44.1kHz) with no way to improve quality beyond what the engine produces
- Users have no control over output sample rate; higher-quality applications (48kHz, cleaner output) require manual post-processing outside QVAC

## 📝 How does it solve it?

Integrates [LavaSR](https://github.com/Topping1/LavaSRcpp), a lightweight neural speech enhancement model, as an opt-in post-processing step with three independent config flags:

- **`enhance`** (bool) — runs Vocos-based neural bandwidth extension to 48kHz (2 ONNX sessions, ~55MB)
- **`denoise`** (bool) — runs UL-UNAS denoiser at 16kHz before enhancement (~1.7MB, 1 ONNX session)
- **`outputSampleRate`** (int) — sets final output rate (must be in 8000–192000 range); with `enhance`, upscales neurally first then resamples conventionally

Pipeline: `Engine synthesize → [denoise at 16kHz] → [enhance to 48kHz] → [resample to target rate]`

All flags default to off — zero performance cost and full backward compatibility.

**C++ implementation:**
- Pure DSP utilities (Lanczos resampler, radix-2 FFT, STFT/ISTFT, Slaney mel filterbank, spectral crossover merge) — no ML dependency, tested in isolation
- LavaSR ONNX wrappers (denoiser with chunked overlap-add, enhancer with backbone + spec head + FastLR merge)
- `TTSModel::postProcess()` integration after engine `synthesize()`, with lazy session loading for per-job toggles

**JS bridge:** `enhance`/`denoise`/`outputSampleRate` + individual model paths (`backbonePath`, `specHeadPath`, `denoiserPath`) flow through `AddonJs.hpp` → `index.js` → `index.d.ts`

**Nice-to-haves included:**
- `sampleRate` exposed in JS output callback (`data.sampleRate`) and `runtimeStats`
- Per-job `enhance`/`denoise`/`outputSampleRate` toggle via `run()` input (lazy ONNX session loading on first use)
- Benchmark tests: enhancer ~22x realtime, denoiser ~48x realtime on Apple Silicon

## 🧪 How was it tested?

**C++ (187 total, 183 pass, 4 skip for missing Supertonic models):**
- 43 new unit tests — DSP round-trip validation (FFT, STFT/ISTFT, resampler, mel filterbank, spectral merge), LavaSR wrapper lifecycle, TTSModel config parsing + resampling + per-job toggle + backward compat
- 6 new integration tests with real ONNX models — enhancer/denoiser individually, full Chatterbox→LavaSR pipeline, denoise+enhance combined, enhance+outputSampleRate
- 3 benchmark tests measuring latency at 1s/3s/5s/10s durations
- Zero regressions on all pre-existing tests

**JS:**
- `outputSampleRate` resampling test added to `addon.test.js` (runs in CI without LavaSR models)
- 6 LavaSR integration tests in `addon.test.js` (Chatterbox+enhance, denoise+enhance, outputSampleRate, enhance+downsample, Supertonic+enhance, backward compat) — all assert `sampleRate` via callback
- Locally verified: outputSampleRate=16000 produces 16kHz output with `sampleRate` correctly reported in callback

**Manual verification:**
- Ran comparison example producing 3 WAV files (raw 24kHz, enhanced 48kHz, denoised+enhanced 48kHz) — enhanced output is audibly clearer

## 🔌 API Changes

```javascript
const tts = new ONNXTTS({
  files: {
    tokenizerPath: 'models/chatterbox/tokenizer.json',
    speechEncoderPath: 'models/chatterbox/speech_encoder.onnx',
    embedTokensPath: 'models/chatterbox/embed_tokens.onnx',
    conditionalDecoderPath: 'models/chatterbox/conditional_decoder.onnx',
    languageModelPath: 'models/chatterbox/language_model.onnx',
  },
  referenceAudio,
  config: {
    language: 'en',
    outputSampleRate: 22050, // optional: resample after enhancement (8000–192000)
  },
  // LavaSR enhancement (nested enhancer config)
  enhancer: {
    type: 'lavasr',
    enhance: true,
    denoise: true,
    backbonePath: 'models/lavasr/enhancer_backbone.onnx',
    specHeadPath: 'models/lavasr/enhancer_spec_head.onnx',
    denoiserPath: 'models/lavasr/denoiser_core_legacy_fixed63.onnx',
  },
})

// Per-job toggle (lazy loads ONNX sessions on first use)
await model.run({
  type: 'text',
  input: 'Hello world',
  enhancer: { type: 'lavasr', enhance: true, denoise: false },
  outputSampleRate: 22050, // per-job override
})

// Output callback now includes sampleRate
response.onUpdate(data => {
  console.log(data.outputArray)  // Int16Array
  console.log(data.sampleRate)   // 48000 (if enhanced)
})
```

New TypeScript interfaces:
```typescript
interface LavaSREnhancerConfig {
  type: 'lavasr'
  enhance?: boolean
  denoise?: boolean
  backbonePath?: string
  specHeadPath?: string
  denoiserPath?: string
}

type EnhancerConfig = LavaSREnhancerConfig

// Added to ONNXTTSOptions:
interface ONNXTTSOptions {
  enhancer?: EnhancerConfig
  // ...existing fields
}

// Added to ONNXTTSRuntimeConfig:
interface ONNXTTSRuntimeConfig {
  outputSampleRate?: number  // 8000–192000
  // ...existing fields
}

// Per-job run input:
interface TTSRunInput {
  enhancer?: { type: 'lavasr'; enhance?: boolean; denoise?: boolean }
  outputSampleRate?: number
  // ...existing fields
}
```

---

## 🔄 CI Run

Latest full CI pipeline: https://github.com/tetherto/qvac/actions/runs/24237672067

### CODE REVIEW

File: packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js [hunk 1]
```diff
@@ -479,6 +479,58 @@ test('Chatterbox Multilingual TTS: WER verification for Spanish', { timeout: 180
...
+      }
+    })
+    .await()
+
+  t.ok(outputArray.length > 0, 'Should produce non-empty output audio')
+
+  console.log(`Output length: ${outputArray.length} samples (resampled from 24kHz to 16kHz)`)
+
+  await model.unload()
+})
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js:372]: Could use supertonic instead of chatterbox since it is faster to run

Also this test is not ensuring that output is 16kHz, only that output is not empty

File: packages/qvac-lib-infer-onnx-tts/index.js [hunk 2]
```diff
@@ -215,10 +268,21 @@ class ONNXTTS extends InferBase {
...
-      })
+      }
+
+      if (input.enhance !== undefined || input.denoise !== undefined || input.outputSampleRate !== undefined) {
+        jobData.config = {}
+        if (input.enhance !== undefined) jobData.config.enhance = input.enhance
+        if (input.denoise !== undefined) jobData.config.denoise = input.denoise
+        if (input.outputSampleRate !== undefined) {
+          jobData.config.outputSampleRate = String(input.outputSampleRate)
+        }
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/index.js:552]: Shouldnt it also register backbonePath, specHeadPath, denoiserPath?

File: packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/TTSModel.cpp [hunk 3]
```diff
@@ -321,4 +346,115 @@ void TTSModel::setReferenceAudio(const std::vector<float> &referenceAudio) {
...
+  auto getString = [&](const std::string &key) -> std::string {
+    auto it = configMap.find(key);
+    return (it != configMap.end()) ? it->second : "";
+  };
+
+  lavaSRConfig_.enhance = getBool("enhance");
+  lavaSRConfig_.denoise = getBool("denoise");
+  lavaSRConfig_.backbonePath = getString("enhancerBackbonePath");
+  lavaSRConfig_.specHeadPath = getString("enhancerSpecHeadPath");
+  lavaSRConfig_.denoiserPath = getString("denoiserPath");
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/TTSModel.cpp]: Shouldnt it only set those if the config is present? So in a reload event the non-specified params are not set to empty values

File: packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/TTSModel.cpp [hunk 4]
```diff
@@ -234,6 +250,13 @@ TTSModel::Output TTSModel::process(const Input &text) {
...
 
+  if (lavaSRConfig_.denoise || lavaSRConfig_.enhance ||
+      lavaSRConfig_.outputSampleRate > 0) {
+    result = postProcess(std::move(result));
+  }
+
+  outputSampleRate_->store(result.sampleRate);
+
   if (cancelRequested_.exchange(false)) {
     throw std::runtime_error("Job cancelled");
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/TTSModel.cpp:311]: Probably is worth to add a cancel check before the postProcess step as well

File: packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/dsp/StftProcessor.cpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  std::vector<float> w(length);
+  for (int i = 0; i < length; i++) {
+    w[i] = 0.5f * (1.0f - std::cos(2.0f * static_cast<float>(PI) * i / length));
+  }
+  return w;
+}
+
+std::vector<float> StftProcessor::padReflect(const std::vector<float> &x,
+                                             int padLeft, int padRight) {
+  const int N = static_cast<int>(x.size());
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/dsp/StftProcessor.cpp:66]: Add defensive guard if (N <= 1) to prevent infinite loop if bogus audio arrives here

File: packages/qvac-lib-infer-onnx-tts/test/integration/lavasr-enhance.test.js [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+  const response = await model.run({ type: 'text', input: 'No enhancement flags.' })
+  const result = await response.await()
+
+  t.ok(result.data.outputArray, 'Should produce output audio')
+  t.ok(result.data.outputArray.length > 0, 'Output should be non-empty')
+
+  console.log(`Output length: ${result.data.outputArray.length} samples (native ${CHATTERBOX_SAMPLE_RATE}Hz, no LavaSR)`)
+
+  await model.unload()
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/test/integration/lavasr-enhance.test.js]: Could you
1 - move these to addon.test.js, it reduces the need to change package.json, mobile tests directory, etc.
2 - add better verification that all enhancements are actually applied and not only output is not empty

File: packages/qvac-lib-infer-onnx-tts/examples/example-enhanced-audio.js
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/examples/example-enhanced-audio.js:1]: Could you change example name to something on the lines of "example-enhanced-audio"?

File: packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/dsp/StftProcessor.cpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  const int numFrames =
+      (static_cast<int>(xpad.size()) - winLength_) / hopLength_ + 1;
+  const int freqBins = nFft_ / 2 + 1;
+
+  Spectrogram spec(numFrames, std::vector<std::complex<float>>(freqBins));
+
+  for (int t = 0; t < numFrames; t++) {
+    ComplexVec frame(nFft_, {0.0f, 0.0f});
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/dsp/StftProcessor.cpp]: frame need not to be created/deleted in every loop, it can be created outside

File: packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/LavaSRDenoiser.cpp [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    }
+    if (starts.back() != T - L) {
+      starts.push_back(T - L);
+    }
+
+    std::vector<float> acc(2 * T * FREQ_BINS, 0.0f);
+    std::vector<float> wacc(T, 0.0f);
+
+    for (int start : starts) {
+      std::vector<float> chunk(2 * L * FREQ_BINS, 0.0f);
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/LavaSRDenoiser.cpp]: chunk need not to be created/deleted in every loop, it can be created outside

File: packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/LavaSRDenoiser.cpp [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      for (int c = 0; c < 2; c++) {
+        for (int t = 0; t < L; t++) {
+          for (int f = 0; f < FREQ_BINS; f++) {
+            chunk[c * L * FREQ_BINS + t * FREQ_BINS + f] =
+                flatSpec[c * T * FREQ_BINS + (start + t) * FREQ_BINS + f];
+          }
+        }
+      }
+
+      std::vector<int64_t> shape = {1, 2, L, FREQ_BINS};
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/LavaSRDenoiser.cpp]: shape need not to be created/deleted in every loop, it can be created outside

File: packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/dsp/MelFilterbank.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+      filters_[i][j] = std::max(0.0f, std::min(lower, upper)) * enorm;
+    }
+  }
+}
+
+std::vector<std::vector<float>>
+MelFilterbank::melSpectrogram(const std::vector<float> &wav,
+                              int hopLength) const {
+  // The enhancer uses center_pad=false (same padding)
+  StftProcessor stft(nFft_, hopLength, nFft_, false);
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/dsp/MelFilterbank.cpp]: Cant we use stft_ member instead of creating it again here?

**Review by GustavoA1604 — CHANGES_REQUESTED** [review]

**Review by ogad-tether — DISMISSED** [review]

File: packages/qvac-lib-infer-onnx-tts/package.json
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/package.json:1]: Please bump to 0.9.0

File: packages/qvac-lib-infer-onnx-tts/CMakeLists.txt [hunk 11]
```diff
@@ -200,6 +206,40 @@ if(ANDROID)
   endif()
 endif()
 
+option(BUILD_CLI "Build CLI tool example" FALSE)
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/CMakeLists.txt]: Why adding option to build_cli in this PR?

File: packages/qvac-lib-infer-onnx-tts/index.js [hunk 12]
```diff
@@ -172,6 +173,25 @@ class ONNXTTS {
       ? lazySessionLoading
       : (platform() === 'ios' || platform() === 'android')
 
+    const outputSampleRate = this._config.outputSampleRate
+    if (outputSampleRate != null && outputSampleRate > 0 && (outputSampleRate < 8000 || outputSampleRate > 192000)) {
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/index.js]: Why check if > 0 as well?

File: packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js [hunk 13]
```diff
@@ -287,6 +287,59 @@ test('Chatterbox Multilingual TTS: Synthesis across languages', { timeout: 36000
...
+  }
+
+  const ONNXTTS = require('../..')
+  const targetRate = 16000
+
+  const model = new ONNXTTS({
+    modelDir,
+    voiceName: 'F1',
+    outputSampleRate: targetRate,
+    opts: { stats: true }
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js:342]: These tests need to be updated to use the new constructor API pattern used by TTS 0.8.0

**Review by GustavoA1604 — CHANGES_REQUESTED** [review]
- Need JS linting
- Also I think automatic on-pr run of this PR got broken, please run manually and share results

File: packages/qvac-lib-infer-onnx-tts/test/integration/lavasr-enhance.test.js (hunk 6 above)
Comment (sharmaraju352) [reply packages/qvac-lib-infer-onnx-tts/test/integration/lavasr-enhance.test.js]: Addressed — tests moved to addon.test.js and now assert sampleRate via callback. All tests use the 0.8.0 constructor API (files:{}, config:{}, enhancer:{}).

File: packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js [hunk 14]
```diff
@@ -528,3 +581,298 @@ test('Supertonic TTS multilingual (Spanish): basic synthesis with HF Supertone/s
...
+
+function loadReferenceAudio () {
+  const { readWavAsFloat32, resampleLinear } = require('../utils/wav-helper')
+  const refPath = path.join(__dirname, '..', 'reference-audio', 'jfk.wav')
+  const { samples, sampleRate } = readWavAsFloat32(refPath)
+  if (sampleRate !== 24000) {
+    return resampleLinear(samples, sampleRate, 24000)
+  }
+  return samples
+}
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js]: These functions can be moved to utils folder

File: packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js [hunk 15]
```diff
@@ -528,3 +581,298 @@ test('Supertonic TTS multilingual (Spanish): basic synthesis with HF Supertone/s
...
+    .await()
+
+  t.ok(result.data.outputArray, 'Should produce output audio')
+  t.ok(result.data.outputArray.length > 0, 'Output should be non-empty')
+  t.is(reportedSampleRate, ENHANCED_SAMPLE_RATE, `Sample rate should be ${ENHANCED_SAMPLE_RATE}Hz after Supertonic enhancement`)
+
+  await model.unload()
+})
+
+test('LavaSR: No flags = backward compatible', { timeout: 1800000 }, async (t) => {
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js]: We need not to add a new test for this. Can simply add the `t.is(reportedSampleRate, 24000, 'Sample rate should remain native 24000Hz without enhancement')` in an existing test (both for supertonic and chatterbox)

File: packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js [hunk 16]
```diff
@@ -528,3 +581,298 @@ test('Supertonic TTS multilingual (Spanish): basic synthesis with HF Supertone/s
...
+  const { readWavAsFloat32, resampleLinear } = require('../utils/wav-helper')
+  const refPath = path.join(__dirname, '..', 'reference-audio', 'jfk.wav')
+  const { samples, sampleRate } = readWavAsFloat32(refPath)
+  if (sampleRate !== 24000) {
+    return resampleLinear(samples, sampleRate, 24000)
+  }
+  return samples
+}
+
+test('LavaSR: Chatterbox + enhance produces 48kHz output', { timeout: 1800000 }, async (t) => {
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js:699]: Could we use supertonic instead of chatterbox in all these tests (except this one) since it runs faster?

File: packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js [hunk 17]
```diff
@@ -287,6 +287,59 @@ test('Chatterbox Multilingual TTS: Synthesis across languages', { timeout: 36000
   console.log('='.repeat(60))
 })
 
+// ---------------------------------------------------------------------------
+// LavaSR outputSampleRate tests (no LavaSR models needed, tests resampling)
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/test/integration/addon.test.js]: This could be simply `outputSampleRate tests` as it is not LavaSR one

**Review by GustavoA1604 — CHANGES_REQUESTED** [review]

File: .github/workflows/on-pr-qvac-lib-infer-onnx-tts.yml [hunk 18]
```diff
@@ -139,7 +139,7 @@ jobs:
     environment: release
     steps:
       - name: Run Sanity checks
-        uses: tetherto/qvac-devops/.github/actions/sanity-checks@v1.1.8
+        uses: tetherto/oss-actions/.github/actions/sanity-checks@v1.1.0
```
Comment (GustavoA1604) [inline .github/workflows/on-pr-qvac-lib-infer-onnx-tts.yml]: Maybe wrong changes?

File: packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/SupertonicEngine.cpp [hunk 19]
```diff
@@ -945,8 +945,7 @@ AudioResult SupertonicEngine::synthesize(const std::string &text) {
   out.channels = 1;
   out.pcm16 = std::move(acc);
   out.samples = out.pcm16.size();
-  out.durationMs =
-      out.samples * 1000.0 / static_cast<double>(sampleRate_);
+  out.durationMs = out.samples * 1000.0 / static_cast<double>(sampleRate_);
```
Comment (GustavoA1604) [inline packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/SupertonicEngine.cpp]: Are these formatting changes expected?

**Review by GustavoA1604 — CHANGES_REQUESTED** [review]

File: .github/workflows/on-pr-qvac-lib-infer-onnx-tts.yml (hunk 18 above)
Comment (sharmaraju352) [reply .github/workflows/on-pr-qvac-lib-infer-onnx-tts.yml]: main already uses oss-actions, after rebase this change is not showing in diff anymore

File: packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/SupertonicEngine.cpp (hunk 19 above)
Comment (sharmaraju352) [reply packages/qvac-lib-infer-onnx-tts/addon/src/model-interface/SupertonicEngine.cpp]: i had ran clang format which caused this, reverted now

**Review by GustavoA1604 — APPROVED** [review]

**Review by moromisato — APPROVED** [review]
